### DISCUSSION TITLE
Implementing Multi-GPU Support in gem5: Modifying mmap to Include GPU ID

### DISCUSSION THREAD

**Discussion by cja1229** [gem5/gem5#1601] [Q&A]
Hello ,
I'm currently working on adding multi-GPU support in gem5, and I've run into a challenge. I need to modify the mmap function in src/gpu-compute/gpu-compute-driver.cc to include the GPU ID.
Could anyone provide guidance on how to implement this? Specifically, I’d appreciate any insights on what changes are necessary to ensure that mmap properly handles GPU IDs for multiple instances.
Thank you for your help!
```
GPUComputeDriver::mmap(ThreadContext *tc, Addr start, uint64_t length,
                       int prot, int tgt_flags, int tgt_fd, off_t offset)
{
    auto process = tc->getProcessPtr();
    auto mem_state = process->memState;

    Addr pg_off = offset >> PAGE_SHIFT;
    Addr mmap_type = pg_off & KFD_MMAP_TYPE_MASK;
    DPRINTF(GPUDriver, "amdkfd mmap (start: %p, length: 0x%x,"
            "offset: 0x%x)\n", start, length, offset);

    switch(mmap_type) {
        case KFD_MMAP_TYPE_DOORBELL:
            DPRINTF(GPUDriver, "amdkfd mmap type DOORBELL offset\n");
            start = mem_state->extendMmap(length);
            process->pTable->map(start, device->hsaPacketProc().pioAddr,
                    length, false);
            break;
        case KFD_MMAP_TYPE_EVENTS:
            DPRINTF(GPUDriver, "amdkfd mmap type EVENTS offset\n");
            panic_if(start != 0,
                     "Start address should be provided by KFD\n");
            panic_if(length != 8 * KFD_SIGNAL_EVENT_LIMIT,
                     "Requested length %d, expected length %d; length "
                     "mismatch\n", length, 8* KFD_SIGNAL_EVENT_LIMIT);
            /**
             * We don't actually access these pages.  We just need to reserve
             * some VA space.  See commit id 5ce8abce for details on how
             * events are currently implemented.
             */
            if (!eventPage) {
                eventPage = mem_state->extendMmap(length);
                start = eventPage;
            }
            break;
        default:
            warn_once("Unrecognized kfd mmap type %llx\n", mmap_type);
            break;
    }

    return start;
}
```

**Comment by mattsinc** [comment]
I'd need to know more about the context here (e.g., why you need GPU ID here) to provide help.  However, my group has recently released our multi-chiplet GPU support (although there is a lot to improve on it) as part of our upcoming MICRO '24 paper.  It still needs to be integrated into the mainline too, so your work may still be valuable.  If you shoot me an email I can explain where that support is at, and we can discuss if what you are doing is better than trying to leverage our stuff.

**Reply by cja1229** [reply]
Thank you for your reply! I found this link : [gem5 Multi-GPU Guide](https://www.gem5.org/2020/05/30/enabling-multi-gpu.html), which explains how a unique GPU ID parameter is added to distinguish each component. However, I’m still a bit confused about how this works in practice.
In the code, particularly in src/gpu-compute/gpu-compute-driver.cc in gem5, it seems to be represented by dGPUPoolID. I appreciate any further insights you might have!

**Reply by mattsinc** [reply]
Yes, that is an article I/we wrote several years ago.  Certainly each chiplet/GPU has a unique ID.  The mainline gem5 code does not have any of that code integrated though, so I would have to look at what dGPUPoolID represents -- it would be a surprise if it was related to this, but I could be wrong.

Regardless as I mentioned above we are working on releasing that support you linked as part of our MICRO paper.  That public support is not completely working yet for reasons I'm still working through, but if your goal is to just get that code you could work with us to verify what is released publicly is useful.  There are downsides though -- for example, that version of the code assumes ROCm 1.6 support, which is not the latest and greatest ROCm version anymore.  And it is on our todo list to get chiplet support for more modern ROCm versions.

This sort of discussion is probably best done via email though.

**Reply by cja1229** [reply]
Thank you for your response!I’ve sent an email regarding my questions about multi-GPU support and the related code. I would greatly appreciate it if you could reply to me.

**Reply by Liujiaqi-jlu** [reply]
> Yes, that is an article I/we wrote several years ago. Certainly each chiplet/GPU has a unique ID. The mainline gem5 code does not have any of that code integrated though, so I would have to look at what dGPUPoolID represents -- it would be a surprise if it was related to this, but I could be wrong.
> 
> Regardless as I mentioned above we are working on releasing that support you linked as part of our MICRO paper. That public support is not completely working yet for reasons I'm still working through, but if your goal is to just get that code you could work with us to verify what is released publicly is useful. There are downsides though -- for example, that version of the code assumes ROCm 1.6 support, which is not the latest and greatest ROCm version anymore. And it is on our todo list to get chiplet support for more modern ROCm versions.
> 
> This sort of discussion is probably best done via email though.

Hello, I would also like to inquire about multi-GPU support. The current version of gem5 does not support instantiating multiple GPUs. I looked at your team's presentation at MICRO 2024, although the paper has not been officially published yet. If I want to implement the simplest multi-GPU interconnect support in FS mode, with only memory semantics communication requirements, is this something feasible? I would like to try modifying gem5. Could you provide some guidance or suggestions? I look forward to your reply!

**Reply by mattsinc** [reply]
You are right that it has not been merged onto the mainline of gem5.  However it is posted publicly (as well as the paper on my website ... not sure why ACM/IEEE is taking so long to post papers this time).  However, note that this support we implemented is only for SE mode, whereas you want FS mode.  So it seems like this is a non-starter?

In any event the artifact is here: https://github.com/hal-uw/cpelide-micro24-artifact/ and the PDF is here: https://pages.cs.wisc.edu/~sinclair/papers/pdalmia-cpelide-micro24.pdf.  For the artifact, there may be some issues as while I thought everything was tested, the AD/AE folks for MICRO had some problems with it.  Let me know if you run into issues.

I don't understand your question about "memory semantics communication requirements" well enough to answer, sorry.

**Reply by Liujiaqi-jlu** [reply]
> You are right that it has not been merged onto the mainline of gem5. However it is posted publicly (as well as the paper on my website ... not sure why ACM/IEEE is taking so long to post papers this time). However, note that this support we implemented is only for SE mode, whereas you want FS mode. So it seems like this is a non-starter?
> 
> In any event the artifact is here: https://github.com/hal-uw/cpelide-micro24-artifact/ and the PDF is here: https://pages.cs.wisc.edu/~sinclair/papers/pdalmia-cpelide-micro24.pdf. For the artifact, there may be some issues as while I thought everything was tested, the AD/AE folks for MICRO had some problems with it. Let me know if you run into issues.
> 
> I don't understand your question about "memory semantics communication requirements" well enough to answer, sorry.

Thank you for your reply. I have read your paper. What we are trying to achieve is an inter-GPU communication experiment similar to NVLink, where we can physically instantiate multiple GPUs for simple communication. In your paper, this could be represented by the communication between multiple MCM GPUs. Based on my initial research, it seems that gem5, whether in SE mode or FS mode, does not support such functionality. Your paper focuses on synchronization within MCM GPUs, which seems somewhat different from what we aim to achieve. Therefore, I would like to ask for your guidance. Is it possible to modify gem5 to implement the functionality we need? If it's difficult to implement, are there other simulators that could offer similar capabilities? I look forward to your guidance!

**Reply by mattsinc** [reply]
sorry, somehow I missed your prior response.  It is in theory possible, but would require that you implement (or mimic) something like NVLink into the simulator.  Currently that support does not exist -- so it would be a lot of work, as you alluded to.

@abmerop has been working on some GPUFS support with disjoint Ruby systems that may make this easier, but I think it would still be a lot of work (but is theoretically plausible).

**Reply by abmerop** [reply]
I do have some uncommitted changes on a branch for stdlib support for multiple GPUs: https://github.com/abmerop/gem5/commits/mgpu/

Currently the status is you can instantiate multiple GPUs and they can run kernels independently. They cannot communicate yet, which I why I have been holding off on posting the commits. If you think it is useful though, I will create a PR.

The GPUs can communicate in two ways: Over PCIe or using XGMI (similar to NVLink for AMD gpus). I would definitely recommend starting with XGMI because compute GPU manufacturers are currently/starting to make compute-only GPUs which do not even have PCIe support.

I spent a little bit of time trying to figure this out but I do not currently have time to work on it. Here are the details of what I know so far:
 - You will at least need to reverse engineer how the MMIO registers *_XGMI_LFB_CNTL and *_XGMI_LFB_SIZE function. There are different names depending on if you are instantiating MI300X or MI210 GPUs in stdlib. Edit: https://elixir.bootlin.com/linux/v6.13/source/drivers/gpu/drm/amd/amdgpu/gfxhub_v2_1.c#L508
 - Once that I setup I *think* the GPUs simply have different physical address ranges and you can just read/write to PIO bus in gem5 to get a minimum viable solution. Realistically it should have its own bus, but you could probably play around with latencies/bandwidth to get something that simulates close enough to an independent bus over PIO.

I'm happy to answer any questions, where possible, but I am not sure when I will be able to get back to this work right now.
